keps/sig-network/3070-reserved-service-ip-range/README.md

```diff
@@ -0,0 +1,751 @@
+type RangeAllocation struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+	// A string representing a unique label for a range of resources, such as a CIDR "10.0.0.0/8" or
+	// port range "10000-30000". Range is not strongly schema'd here. The Range is expected to define
+	// a start and end unless there is an implicit end.
+	Range string
+	// A byte array representing the serialized state of a range allocation. Additional clarifiers on
+	// the type or format of data should be represented with annotations. For IP allocations, this is
+	// represented as a bit array starting at the base IP of the CIDR in Range, with each bit representing
+	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
+	Data []byte
+}
+```
+
+The [apiserver Service registry contains allocators backed by those bitmaps](p[kg/registry/core/rest/storage_core.go](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/pkg/registry/core/rest/storage_core.go#L96-L103)) in order to assign these IPs.
+
+```go
+type LegacyRESTStorage struct {
+	ServiceClusterIPAllocator          rangeallocation.RangeRegistry
+	SecondaryServiceClusterIPAllocator rangeallocation.RangeRegistry
+	ServiceNodePortAllocator           rangeallocation.RangeRegistry
+}
+
+...
+func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generic.RESTOptionsGetter) (LegacyRESTStorage, genericapiserver.APIGroupInfo, error) {
+  ...
+	serviceClusterIPAllocator, err := ipallocator.New(&serviceClusterIPRange, func(max int, rangeSpec string) (allocator.Interface, error) {
+		mem := allocator.NewAllocationMap(max, rangeSpec)
+		// TODO etcdallocator package to return a storage interface via the storageFactory
+		etcd, err := serviceallocator.NewEtcd(mem, "/ranges/serviceips", serviceStorageConfig.ForResource(api.Resource("serviceipallocations")))
+		if err != nil {
+			return nil, err
+		}
+		serviceClusterIPRegistry = etcd
+		return etcd, nil
+	})
+```
+
+The bitmaps implement a [strategy for allocating a random free value if none is specified](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/pkg/registry/core/service/allocator/bitmap.go#L186-L205):
+
+```go
+// randomScanStrategy chooses a random address from the provided big.Int, and then
+// scans forward looking for the next available address (it will wrap the range if
+// necessary).
+type randomScanStrategy struct {
+	rand *rand.Rand
+}
+
+func (rss randomScanStrategy) AllocateBit(allocated *big.Int, max, count int) (int, bool) {
+	if count >= max {
+		return 0, false
+	}
+	offset := rss.rand.Intn(max)
+	for i := 0; i < max; i++ {
+		at := (offset + i) % max
+		if allocated.Bit(at) == 0 {
+			return at, true
+		}
+	}
+	return 0, false
+}
+```
+
+The proposal is to implement a new strategy on the Service IP allocation bitmap that avoids to use the first 32 values
+if there are free values on the upper block of the range.
```

Comment:
That seems weird. I would expect that if there was a reserved range, that it would be reserved unconditionally, not "reserved until we run out of unreserved addresses".

Also, it's weird to hardcode the reserved range to 32 addresses regardless of the size of the service CIDR...

Reply:
Heh.  We went back and forth on this a lot in discussing.

If your service CIDR is /27, you only get 32 service IPs (IIRC, it's 30 actually, we avoid the first and last).  At that point we can debate whether it's OK to have so few and if, at that point, we can make many guarantees.

If your CIDR is bigger than that, do we need to reserve more than 32 IPs?  If so, do we need a "cascade" of blocks?

e.g. something like `min(32, cidrSize/2)` is attractive, but only useful if we think there are a lot of users with really small ranges.

Something more dynamic like "always prefer the upper half unless it is full" is more general, I guess.  Is it worth the extra effort?

E.g. assume a /22 service range (1024 IPs).   First look at [512]-[1023].  If that's full, look at [256-511].  If that's full, look at [128-255].  It seems like that gives the caller influence where it isn't needed, which could be a vector for some sort of attack.

`min(max(16, cidrSize/16), 256)` ?  At /28 it would be 16 (full range).  At /27 - /24 it would be 16 (part of the range).  At /23 it would be 32.  /22 -> 64. /21 ->128.  /20 and bigger -> 256.

Again, worth the effort?

Reply:
what about making it a fixed percentage? we subdivide in two ranges, 

```





           ┌─────────────┬─────────────────────────────────────────────┐
           │             │                                             │
           │             │                                             │
           │             │                                             │
           │   static    │                    dynamic                  │
           │             │                                             │
           │             │                                             │
           │             │                                             │
           └─────────────┴─────────────────────────────────────────────┘

           ◄────────────► ◄────────────────────────────────────────────►

                15%                              85%
```

Reply:
"15%" must be a typo, since it is not a power of two.  I think you meant to type "12.5%" :)

Seriously though, I think percents are OK for humans to get a big-picture, but harder to be precise about.

`min(max($min, cidrSize/$step), $max)` is easy to describe as "never less than $min or more than $max, with a graduated step function between them.  We can argue about values for those variables (or even tweak them over time if we get feedback), but the logic still seems appropriate?

I proposed:

$min = 16
$max = 256
$step = 16

But I am fine to haggle :)

Reply:
I'm fine with anything meanwhile users can understand it

Reply:
updated with the suggested formula

Reply:
FTR I meant that 32 seemed like too many reserved IPs for small service CIDRs, not that I thought we should reserve more for large CIDRs.

But I guess it's reasonable to assume that users with larger clusters are likely to have more use cases for fixed-IP services, so I guess giving them a larger range makes sense. And if we want the reserved size to (a) always be a power of 2, and (b) always include the .10 address, then I guess 16 has to be the minimum value. And given that the "reserved" range is more like a "preferred" range, then even 16 isn't too large for a /27